zcash-issue-dag.py

```diff
@@ -298,23 +303,74 @@ class Meta:
 
     def __init__(self, *args, **kw):
         super(ZenHubAPI, self).__init__(*args, **kw)
-        self.add_resource('dependencies', ZHDepsResourceHandler)
         self.add_resource('epics', ZHEpicsResourceHandler)
         self.add_resource('epics_issues', ZHEpicsIssuesResourceHandler)
 
     def auth(self, *args, **kw):
         pass
 
+def fetch_workspace_graph(op, workspace_id, repos, cursor):
+    dependencies = op.workspace(id=workspace_id).issue_dependencies(
+        # TODO: This causes a 500 Internal Server Error.
+        # repository_ids=repos,
+        first=100,
+        after=cursor,
+    )
+    dependencies.nodes.id()
+    dependencies.nodes.blocked_issue.number()
+    dependencies.nodes.blocked_issue.repository.gh_id()
+    dependencies.nodes.blocking_issue.number()
+    dependencies.nodes.blocking_issue.repository.gh_id()
+    dependencies.page_info.has_next_page()
+    dependencies.page_info.end_cursor()
+
+def get_dependency_graph(endpoint, workspace_id, repos):
+    edges = []
+    cursor = None
+
+    while True:
+        op = Operation(zenhub_schema.Query)
+        fetch_workspace_graph(op, workspace_id, repos, cursor)
+
+        d = endpoint(op)
+        data = (op + d)
+
+        dependencies = data.workspace.issue_dependencies
+        edges += [
+            (
+                (node.blocking_issue.repository.gh_id, node.blocking_issue.number),
+                (node.blocked_issue.repository.gh_id, node.blocked_issue.number),
+            ) for node in dependencies.nodes
+        ]
+
+        if dependencies.page_info.has_next_page:
+            cursor = dependencies.page_info.end_cursor
+            print('.', end='', flush=True)
+        else:
+            print()
+            break
+
+    return nx.DiGraph(edges)
 
 def main():
     gapi = HTTPEndpoint(
         'https://api.github.com/graphql',
         {'Authorization': 'bearer %s' % GITHUB_TOKEN},
     )
-    zapi = ZenHubAPI()
+    zapi = HTTPEndpoint(
+        'https://api.zenhub.com/public/graphql',
+        {'Authorization': 'Bearer %s' % ZENHUB_TOKEN},
+    )
+
+    # Build the full dependency graph from ZenHub's per-workspace API.
+    print('Fetching graph')
+    dg = nx.compose_all([
+        get_dependency_graph(zapi, workspace_id, repos)
+        for (workspace_id, repos) in WORKSPACES.items()
+        if len(repos) > 0
+    ])
 
-    # Build the full dependency graph from ZenHub's per-repo APIs.
-    dg = nx.compose_all([zapi.dependencies.get(x) for x in REPOS])
+    print('Rendering DAG')
 
     if SHOW_EPICS:
```

Comment:
This is broken by the PR, as `zapi` no longer points to the REST API, and the epics logic has not yet been rewritten to GraphQL.

Reply:
Can we just strip out epics support? We haven't used them for quite a while. If there is any significant part of the graph still using epics then we can convert it to use plain issues.

Reply:
I've just about finished migrating epics over 😅

Reply:
Damn, and there was me looking for an excuse to get rid of epics completely 😅 (I find the interface awkward and they don't do anything we don't get from target issues and milestones.)